### ISSUE TITLE
finish experimental response post-processing

### ISSUE THREAD

**Issue by Geal** [apollographql/router#100]
PR #86 reduces the performance when processing queries. In a benchmark, we saw that the breaking point went from 15k rps to 3k rps with that PR.

For now, #99 feature gates the response post processing, but this is a temporary solution, we need a proper fix soon.

For more context about the performance issue, here's a flamegraph of the `format_response` function:
![Screenshot from 2021-11-08 17-04-05](https://user-images.githubusercontent.com/119296/140776517-d73a3c45-f91c-494e-b57c-25914f5090e7.png)

That function takes about 3% of the total processing time, of which a large part happens in parsing the query and extracting selection sets (about 2.7% of the total processing time). The call to `apply_selection_set` that actually modifies the response takes about 0.22%, which is fine for now.

We might want to cache the result of the query parsing, but the selection set structure comes from apollo-rs, and contains references to the input data.
So caching that would require storing a copy of the query, and the apollo-rs `Document` that refers to it, and hope that we don't get too many issues with self referential structures.
Another path would be to have a completely owned structure to represent selection sets. The structure would not be too complex for now, but I worry this will collide with future work about query planning and execution.

Any thoughts on this?

**Comment by BrynCooke** [comment]
Caching is likely the primary way that we will want to solve this. The frequency that completely new queries need to be parsed is going to be very small in a production system, with the exception that a user is following an anti-pattern of embedding variables directly in their query body (this can be detected and we can give the user feedback).

Any optimisation that we do will unlikely be as fast as using a cache.

Strong minus 1 on trying a different parser.

**Comment by BrynCooke** [comment]
Also caching is required for APQ: https://www.apollographql.com/docs/apollo-server/performance/apq/
We have to support this.

**Comment by abernix** [comment]
Since you mentioned APQ, I've opened https://github.com/apollographql/router/issues/130 to track that conversation.   (It's worth noting that APQ is typically a `<hash: String, query: String>` map — or that it's on a network store — and that there's a different cache to solve `<query: String, document: GraphqlDocument>`)

**Comment by cecton** [comment]
> The frequency that completely new queries need to be parsed is going to be very small in a production system, with the exception that a user is following an anti-pattern of embedding variables directly in their query body

Ah right makes sense! I didn't know about that tbh, my experience in using GraphQL is rather limited. Thanks for the info!

> Strong minus 1 on trying a different parser.

Yes it wasn't on the top of my list but how hard could it be to parse `{ name products { review } }`? As I said, it's possible to go minimalist on this just to check if the parser itself is not at fault. I tried @Geal's `nom` on a personal project and it was very easy! We could simply make a test crate that parses the exact same thing, very limited, and make a bench test and see how it compares. Maybe that has been done already?

**Another important related point:**

@Geal did you notice how the parsers' nodes are `!Send`? This might be an issue for caching as we would need to transform the parsed document to another representation of our own. This is not a problem per se but afaik it might conflict with @lrlna's TODO list: she's the one in charge to make a higher level representation of the parsed document that we will be able to use here in router.

The problem is: if we do in `router` our own structs for this, it's kinda duplicated work with her higher level API. I would also personally prefer that we **don't** add new "models" in `router` itself for that. This should live in `apollo-rs`.

Anyway! We're not in a hurry anymore so I think we can take our time and do this properly. I personally also want to refactor some stuffs here and there before going to far in the graphql features.

**Comment by Geal** [comment]
I'd prefer we use apollo-rs rather than maintain two different graphql parsers, all projects will benefit from any optimizations we can do, and we would avoid issues with parsers drifting away from each other and accepting different things. If it is accepted or generated by rover, it must work in the router.

> @Geal did you notice how the parsers' nodes are !Send? 

I noticed, that's why I added more owned structures to represent the selections we want to see. I do not see that as a conflict with @lrlna's roadmap, it's a way to leverage apollo-rs right now without putting pressure on its development. We can already try things and provide use cases, having them work right now, then replace them with the proper API once it is done. Code is cheap, if it can solve a problem right now and be deleted in the future, I'm totally fine with it.

**Comment by lrlna** [comment]
@Geal @cecton if there are specific accessors you'd like apollo-rs to have, do let me know! I obviously don't want you to have to wait for me to have some sort of higher level representation to move your work forward, but I am definitely keen to know if you require something specific. That way I can account for it early on in the design process.

**Comment by Geal** [comment]
@lrlna sure, I can gather the list of what we're using right now

**Comment by cecton** [comment]
> I'd prefer we use apollo-rs rather than maintain two different graphql parsers, all projects will benefit from any optimizations we can do, and we would avoid issues with parsers drifting away from each other and accepting different things. If it is accepted or generated by rover, it must work in the router.

I never said making 2 different parsers :confounded: I said: asserting the speed of different parsers.

**Comment by Geal** [comment]
We can now update apollo-rs to this commit and parse queries faster https://github.com/apollographql/apollo-rs/commit/6e8e8a8b23c5489fdb2154029488dfcfe9cba80a

**Comment by cecton** [comment]
Another flamegraph would be nice :innocent:

**Comment by cecton** [comment]
Assigned myself for #106 and @Geal for the perf optimizations on the parser itself.

**Comment by cecton** [comment]
Cross-linking PR https://github.com/apollographql/apollo-rs/pull/113

**Comment by cecton** [comment]
Closed by mistake

**Comment by Geal** [comment]
🥳
